TESTS/mbedmicro-rtos-mbed/semaphore/main.cpp

```diff
@@ -57,7 +67,164 @@ int main (void) {
             break;
         }
     }
+}
+
+struct thread_data {
+    Semaphore *sem;
+    uint32_t data;
+};
+
+void single_thread(struct thread_data *data)
+{
+    int32_t cnt = data->sem->wait();
+    TEST_ASSERT_EQUAL(1, cnt);
+    data->data++;
+}
+
+/** Test single thread
+
+    Given a two threads A & B and a semaphore (with count of 0) and a counter (equals to 0)
+    when thread B calls @a wait
+    then thread B waits for a token to become available
+    then the counter is equal to 0
+    when thread A calls @a release on the semaphore
+    then thread B acquires a token and increments the counter
+    then the counter equals to 1
+ */
+void test_single_thread()
+{
+    Thread t(osPriorityNormal, THREAD_STACK_SIZE);
+    Semaphore sem(0);
+    struct thread_data data;
+    osStatus res;
+
+    data.sem = &sem;
+    data.data = 0;
+
+    res = t.start(callback(single_thread, &data));
+    TEST_ASSERT_EQUAL(osOK, res);
+    Thread::yield();
```

Comment:
you might consider using wait instead of yield here to make this test more deterministic. Depending when in the tick you yield the new thread still might not have enough time to call semaphore wait. Also, if the priority of main is below the priority of the new thread, then the yield won't help.